### PR TITLE
Add vscode on drac

### PR DESCRIPTION
**Pull request by satyaog**
VS Code in web browser on DRAC cluster

### CODE REVIEW

**Comment by satyaog** [comment]
With the following code in `~/.lmod.d/default` VSCode automatically opens after the allocation is given. The downside is that the modules are automatically loaded which might conflict with other modules that could be loaded in other contexts

```lua
-- -*- lua -*-
-- created: Wed Dec 13 11:28:26 2023 --
-- Lmod 8.7.4  2022-06-07 18:23 -05:00
_ModuleTable_ = {
  MTversion = 3,
  c_rebuildTime = false,
  c_shortTime = false,
  depthT = {},
  family = {
    base_os = "gentoo",
    compiler = "intel",
    mpi = "openmpi",
  },
  mT = {
    CCconfig = {
      fn = "/cvmfs/soft.computecanada.ca/custom/modules/CCconfig.lua",
      fullName = "CCconfig",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 1,
      propT = {},
      stackDepth = 1,
      status = "active",
      userName = "CCconfig",
      wV = "M.*zfinal",
    },
    StdEnv = {
      fn = "/cvmfs/soft.computecanada.ca/custom/modules/StdEnv/2020.lua",
      fullName = "StdEnv/2020",
      hash = "2c4a5307fd202186045e0ac470f4c149d97586b4",
      loadOrder = 9,
      propT = {
        lmod = {
          sticky = 1,
        },
      },
      stackDepth = 0,
      status = "active",
      userName = "StdEnv",
      wV = "s00002020.*zfinal",
    },
    ["code-server"] = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/code-server/3.5.0.lua",
      fullName = "code-server/3.5.0",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 15,
      propT = {},
      stackDepth = 0,
      status = "active",
      userName = "code-server/3.5.0",
      wV = "000000003.000000005.*zfinal",
    },
    gcccore = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/gcccore/.9.3.0.lua",
      fullName = "gcccore/.9.3.0",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 3,
      propT = {},
      ref_count = 1,
      stackDepth = 2,
      status = "active",
      userName = "gcccore/.9.3.0",
      wV = "000000000.000000009.000000003.*zfinal",
    },
    gentoo = {
      fn = "/cvmfs/soft.computecanada.ca/custom/modules/gentoo/2020.lua",
      fullName = "gentoo/2020",
      hash = "d9fd95e7e0c4817223fbc3046657d64294c40b41",
      loadOrder = 2,
      propT = {
        lmod = {
          sticky = 1,
        },
      },
      stackDepth = 1,
      status = "active",
      userName = "gentoo/2020",
      wV = "^00002020.*zfinal",
    },
    imkl = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/imkl/2020.1.217.lua",
      fullName = "imkl/2020.1.217",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 4,
      propT = {
        type_ = {
          math = 1,
        },
      },
      stackDepth = 2,
      status = "active",
      userName = "imkl/2020.1.217",
      wV = "^00002020.000000001.000000217.*zfinal",
    },
    intel = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/intel/2020.1.217.lua",
      fullName = "intel/2020.1.217",
      hash = "ded4276fd94089715b6aa921a1491d1c0dc6d3f5",
      loadOrder = 5,
      propT = {
        type_ = {
          tools = 1,
        },
      },
      stackDepth = 1,
      status = "active",
      userName = "intel",
      wV = "^00002020.000000001.000000217.*zfinal",
    },
    ipykernel = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/ipykernel/2023b.lua",
      fullName = "ipykernel/2023b",
      hash = "df4278d9f3701d66e735c1fe4c386a65688e6d35",
      loadOrder = 13,
      propT = {},
      ref_count = 1,
      stackDepth = 1,
      status = "active",
      userName = "ipykernel/2023b",
      wV = "000002023.*b.*zfinal",
    },
    ["ipython-kernel"] = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/ipython-kernel/3.9.lua",
      fullName = "ipython-kernel/3.9",
      hash = "4025ec4661c054a0481e7a1cd9a3b212ac0870c5",
      loadOrder = 14,
      propT = {
        lmod = {
          sticky = 1,
        },
      },
      stackDepth = 0,
      status = "active",
      userName = "ipython-kernel/3.9",
      wV = "000000003.000000009.*zfinal",
    },
    libfabric = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/libfabric/1.10.1.lua",
      fullName = "libfabric/1.10.1",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 7,
      propT = {},
      ref_count = 1,
      stackDepth = 2,
      status = "active",
      userName = "libfabric/1.10.1",
      wV = "000000001.000000010.000000001.*zfinal",
    },
    libffi = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/libffi/3.3.lua",
      fullName = "libffi/3.3",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 11,
      propT = {},
      ref_count = 1,
      stackDepth = 2,
      status = "active",
      userName = "libffi/3.3",
      wV = "000000003.000000003.*zfinal",
    },
    mii = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/mii/1.1.2.lua",
      fullName = "mii/1.1.2",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 10,
      propT = {},
      stackDepth = 0,
      status = "active",
      userName = "mii",
      wV = "000000001.000000001.000000002.*zfinal",
    },
    openmpi = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Compiler/intel2020/openmpi/4.0.3.lua",
      fullName = "openmpi/4.0.3",
      hash = "f05dea3c5ca4eea41b5d747410086f5a416b88fc",
      loadOrder = 8,
      propT = {
        type_ = {
          mpi = 1,
        },
      },
      stackDepth = 1,
      status = "active",
      userName = "openmpi",
      wV = "000000004.000000000.000000003.*zfinal",
    },
    python = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/python/3.9.6.lua",
      fullName = "python/3.9.6",
      hash = "73db2114442f0d2f054f7538f82c47881b40db52",
      loadOrder = 12,
      propT = {
        type_ = {
          tools = 1,
        },
      },
      ref_count = 2,
      stackDepth = 1,
      status = "active",
      userName = "python/3.9.6",
      wV = "000000003.000000009.000000006.*zfinal",
    },
    ucx = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/ucx/1.8.0.lua",
      fullName = "ucx/1.8.0",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 6,
      propT = {},
      ref_count = 1,
      stackDepth = 2,
      status = "active",
      userName = "ucx/1.8.0",
      wV = "000000001.000000008.*zfinal",
    },
  },
  mpathA = {
    "/opt/software/modulefiles"
    , "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/MPI/intel2020/openmpi4"
    , "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Compiler/intel2020"
    , "/home/sogagne/.local/easybuild/modules/2020/avx2/Core"
    , "/home/sogagne/.local/easybuild/modules/2020/Core"
    , "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core"
    , "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core", "/cvmfs/soft.computecanada.ca/custom/modules",
  },
  mpathRefCountT = {
    ["/cvmfs/soft.computecanada.ca/custom/modules"] = 1,
    ["/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core"] = 1,
    ["/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Compiler/intel2020"] = 1,
    ["/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core"] = 1,
    ["/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/MPI/intel2020/openmpi4"] = 1,
    ["/home/sogagne/.local/easybuild/modules/2020/Core"] = 1,
    ["/home/sogagne/.local/easybuild/modules/2020/avx2/Core"] = 1,
    ["/opt/software/modulefiles"] = 1,
  },
  systemBaseMPATH = "/cvmfs/soft.computecanada.ca/custom/modules",
}

```

File: milatools/cli/commands.py [hunk 1]
```diff
@@ -638,6 +663,29 @@ def code(
         print(T.bold(f"  ssh mila scancel {data['jobid']}"))
 
 
+def code_drac(cluster:str, account:str, time:float, cpus:int, mem:int, gpu:str):
+    """Open a remote VSCode session on a drac compute node."""
+    url = f"https://jupyterhub.{cluster}.computecanada.ca"
+    _user="sogagne"
```
Comment (satyaog) [inline milatools/cli/commands.py:669]: The line `_user="sogagne"` is only to test with my user

**Comment by satyaog** [comment]
With the following code in `~/.lmod.d/default` VSCode automatically opens after the allocation is given. The downside is that the modules are automatically loaded which might conflict with other modules that could be loaded in other contexts. The file has been generated within the jupyterlab module interface.

To test, run the following: `mila code-drac cedar --account rrg-bengioy-ad_cpu --time 1 --cpus 1 --mem 1000`

```lua
-- -*- lua -*-
-- created: Wed Dec 13 11:28:26 2023 --
-- Lmod 8.7.4  2022-06-07 18:23 -05:00
_ModuleTable_ = {
  MTversion = 3,
  c_rebuildTime = false,
  c_shortTime = false,
  depthT = {},
  family = {
    base_os = "gentoo",
    compiler = "intel",
    mpi = "openmpi",
  },
  mT = {
    CCconfig = {
      fn = "/cvmfs/soft.computecanada.ca/custom/modules/CCconfig.lua",
      fullName = "CCconfig",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 1,
      propT = {},
      stackDepth = 1,
      status = "active",
      userName = "CCconfig",
      wV = "M.*zfinal",
    },
    StdEnv = {
      fn = "/cvmfs/soft.computecanada.ca/custom/modules/StdEnv/2020.lua",
      fullName = "StdEnv/2020",
      hash = "2c4a5307fd202186045e0ac470f4c149d97586b4",
      loadOrder = 9,
      propT = {
        lmod = {
          sticky = 1,
        },
      },
      stackDepth = 0,
      status = "active",
      userName = "StdEnv",
      wV = "s00002020.*zfinal",
    },
    ["code-server"] = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/code-server/3.5.0.lua",
      fullName = "code-server/3.5.0",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 15,
      propT = {},
      stackDepth = 0,
      status = "active",
      userName = "code-server/3.5.0",
      wV = "000000003.000000005.*zfinal",
    },
    gcccore = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/gcccore/.9.3.0.lua",
      fullName = "gcccore/.9.3.0",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 3,
      propT = {},
      ref_count = 1,
      stackDepth = 2,
      status = "active",
      userName = "gcccore/.9.3.0",
      wV = "000000000.000000009.000000003.*zfinal",
    },
    gentoo = {
      fn = "/cvmfs/soft.computecanada.ca/custom/modules/gentoo/2020.lua",
      fullName = "gentoo/2020",
      hash = "d9fd95e7e0c4817223fbc3046657d64294c40b41",
      loadOrder = 2,
      propT = {
        lmod = {
          sticky = 1,
        },
      },
      stackDepth = 1,
      status = "active",
      userName = "gentoo/2020",
      wV = "^00002020.*zfinal",
    },
    imkl = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/imkl/2020.1.217.lua",
      fullName = "imkl/2020.1.217",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 4,
      propT = {
        type_ = {
          math = 1,
        },
      },
      stackDepth = 2,
      status = "active",
      userName = "imkl/2020.1.217",
      wV = "^00002020.000000001.000000217.*zfinal",
    },
    intel = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/intel/2020.1.217.lua",
      fullName = "intel/2020.1.217",
      hash = "ded4276fd94089715b6aa921a1491d1c0dc6d3f5",
      loadOrder = 5,
      propT = {
        type_ = {
          tools = 1,
        },
      },
      stackDepth = 1,
      status = "active",
      userName = "intel",
      wV = "^00002020.000000001.000000217.*zfinal",
    },
    ipykernel = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/ipykernel/2023b.lua",
      fullName = "ipykernel/2023b",
      hash = "df4278d9f3701d66e735c1fe4c386a65688e6d35",
      loadOrder = 13,
      propT = {},
      ref_count = 1,
      stackDepth = 1,
      status = "active",
      userName = "ipykernel/2023b",
      wV = "000002023.*b.*zfinal",
    },
    ["ipython-kernel"] = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/ipython-kernel/3.9.lua",
      fullName = "ipython-kernel/3.9",
      hash = "4025ec4661c054a0481e7a1cd9a3b212ac0870c5",
      loadOrder = 14,
      propT = {
        lmod = {
          sticky = 1,
        },
      },
      stackDepth = 0,
      status = "active",
      userName = "ipython-kernel/3.9",
      wV = "000000003.000000009.*zfinal",
    },
    libfabric = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/libfabric/1.10.1.lua",
      fullName = "libfabric/1.10.1",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 7,
      propT = {},
      ref_count = 1,
      stackDepth = 2,
      status = "active",
      userName = "libfabric/1.10.1",
      wV = "000000001.000000010.000000001.*zfinal",
    },
    libffi = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/libffi/3.3.lua",
      fullName = "libffi/3.3",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 11,
      propT = {},
      ref_count = 1,
      stackDepth = 2,
      status = "active",
      userName = "libffi/3.3",
      wV = "000000003.000000003.*zfinal",
    },
    mii = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core/mii/1.1.2.lua",
      fullName = "mii/1.1.2",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 10,
      propT = {},
      stackDepth = 0,
      status = "active",
      userName = "mii",
      wV = "000000001.000000001.000000002.*zfinal",
    },
    openmpi = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Compiler/intel2020/openmpi/4.0.3.lua",
      fullName = "openmpi/4.0.3",
      hash = "f05dea3c5ca4eea41b5d747410086f5a416b88fc",
      loadOrder = 8,
      propT = {
        type_ = {
          mpi = 1,
        },
      },
      stackDepth = 1,
      status = "active",
      userName = "openmpi",
      wV = "000000004.000000000.000000003.*zfinal",
    },
    python = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/python/3.9.6.lua",
      fullName = "python/3.9.6",
      hash = "73db2114442f0d2f054f7538f82c47881b40db52",
      loadOrder = 12,
      propT = {
        type_ = {
          tools = 1,
        },
      },
      ref_count = 2,
      stackDepth = 1,
      status = "active",
      userName = "python/3.9.6",
      wV = "000000003.000000009.000000006.*zfinal",
    },
    ucx = {
      fn = "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core/ucx/1.8.0.lua",
      fullName = "ucx/1.8.0",
      hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709",
      loadOrder = 6,
      propT = {},
      ref_count = 1,
      stackDepth = 2,
      status = "active",
      userName = "ucx/1.8.0",
      wV = "000000001.000000008.*zfinal",
    },
  },
  mpathA = {
    "/opt/software/modulefiles"
    , "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/MPI/intel2020/openmpi4"
    , "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Compiler/intel2020"
    , "/home/sogagne/.local/easybuild/modules/2020/avx2/Core"
    , "/home/sogagne/.local/easybuild/modules/2020/Core"
    , "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core"
    , "/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core", "/cvmfs/soft.computecanada.ca/custom/modules",
  },
  mpathRefCountT = {
    ["/cvmfs/soft.computecanada.ca/custom/modules"] = 1,
    ["/cvmfs/soft.computecanada.ca/easybuild/modules/2020/Core"] = 1,
    ["/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Compiler/intel2020"] = 1,
    ["/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/Core"] = 1,
    ["/cvmfs/soft.computecanada.ca/easybuild/modules/2020/avx2/MPI/intel2020/openmpi4"] = 1,
    ["/home/sogagne/.local/easybuild/modules/2020/Core"] = 1,
    ["/home/sogagne/.local/easybuild/modules/2020/avx2/Core"] = 1,
    ["/opt/software/modulefiles"] = 1,
  },
  systemBaseMPATH = "/cvmfs/soft.computecanada.ca/custom/modules",
}
```

**Comment by lebrice** [comment]
Does this open vscode in the browser? Or is it an app link that opens the vscode client?

**Comment by lebrice** [comment]
I'll check about the module load, if we can keep everything in the same command, with the same syntax, that would definitely be best.

**Comment by satyaog** [comment]
I'm in favour of any solution that is not time consuming for a first version, then build or take the time to rebuild properly afterwards. I don't think we have to see the web solution as a final version, discrediting other solutions. A different API for a plug and play solution might not be too bad a trade off. There's a possibility we failed a bit at communicating that DRAC already offered a solution for vs code. We could even aim at a first that could only be an advertisement for that solution, while we build something better in the meantime

**Comment by lebrice** [comment]
I don't seem to be able to sign in with my GitHub account in the web browser VsCode. Do you think that's intended or a bug?
It's very useful to synchronize all the editor settings (extensions, keyboard shortcuts, etc).

I'm currently testing out if adding the code-server module could help with installing extensions on the compute node. So far, it doesn't look like it. I'm gonna try to understand what they do with jupyterlab to make this work.

**Pull request closed without merging**